File: leetcode/algorithms/463_island_perimeter/main.py
from typing import List


class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        result = 0

        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == 0:
                    continue

                result += 4

                if j > 0 and grid[i][j - 1] == 1:
                    result -= 2
                if i > 0 and grid[i - 1][j] == 1:
                    result -= 2

        return result


    # Best Solution
    def bestSolution(self, grid: List[List[int]]) -> int:
        m, n, Perimeter = len(grid), len(grid[0]), 0

        for i in range(m):
            for j in range(n):
                Perimeter += 4 * grid[i][j]
                if i > 0:
                    Perimeter -= grid[i][j] * grid[i - 1][j]
                if i < m - 1:
                    Perimeter -= grid[i][j] * grid[i + 1][j]
                if j > 0:
                    Perimeter -= grid[i][j] * grid[i][j - 1]
                if j < n - 1:
                    Perimeter -= grid[i][j] * grid[i][j + 1]

        return Perimeter

File: leetcode/algorithms/463_island_perimeter/main.js
/**
 * @param {number[][]} grid
 * @return {number}
 */
var islandPerimeter = function (grid) {
  let result = 0;

  for (let i = 0; i < grid.length; i++) {
    for (let j = 0; j < grid[i].length; j++) {
      if (!grid[i][j]) {
        continue;
      }

      result += 4;

      if (j > 0 && grid[i][j - 1]) {
        result -= 2;
      }
      if (i > 0 && grid[i - 1][j]) {
        result -= 2;
      }
    }
  }

  return result;
};


// Solution
var solution = function (grid) {
  const rows = grid.length;
  const cols = grid[0].length;

  var perimeter = 0;

  for (var row = 0; row < rows; row++) {
    for (var col = 0; col < cols; col++) {
      if (!grid[row][col]) continue;

      perimeter += 4;

      // abstract the number of adjacent island
      if (row > 0 && grid[row - 1][col]) perimeter--;
      if (col > 0 && grid[row][col - 1]) perimeter--;
      if (row < rows - 1 && grid[row + 1][col]) perimeter--;
      if (col < cols - 1 && grid[row][col + 1]) perimeter--;
    }
  }

  return perimeter;
};
